unity/src/main/java/io/delta/unity/UCCatalogManagedCommitter.java

```diff
@@ -0,0 +1,196 @@
+  }
+
+  ////////////////////
+  // Helper methods //
+  ////////////////////
+
+  private String normalize(Path path) {
+    return path.toUri().normalize().toString();
+  }
+
+  private void validateCommitMetadataLogPath(CommitMetadata cm) {
+    final String expectedDeltaLogPathNormalized = normalize(new Path(tablePath, "_delta_log"));
+    final String providedDeltaLogPathNormalized = normalize(new Path(cm.getDeltaLogDirPath()));
+    checkArgument(
+        expectedDeltaLogPathNormalized.equals(providedDeltaLogPathNormalized),
+        "Delta log path '%s' does not match expected '%s'",
+        expectedDeltaLogPathNormalized,
+        providedDeltaLogPathNormalized);
+  }
+
+  private FileStatus writeStagedCommitFile(
+      Engine engine, CloseableIterator<Row> finalizedActions, CommitMetadata commitMetadata)
+      throws CommitFailedException {
+    final String stagedCommitFilePath = commitMetadata.getNewStagedCommitFilePath();
+
+    return OperationTimer.timeOperation(
+        logger,
+        "Write staged commit file",
+        ucTableId,
+        () -> {
+          try {
+            // Do not use Put-If-Absent for staged commit files since we assume that UUID-based
+            // commit files are globally unique, and so we will never have concurrent writers
+            // attempting to write the same commit file.
+
+            engine
+                .getJsonHandler()
+                .writeJsonFileAtomically(
+                    stagedCommitFilePath, finalizedActions, true /* overwrite */);
+
+            // TODO: [delta-io/delta#5021] Use FileSystemClient::getFileStatus API instead
+            return FileStatus.of(stagedCommitFilePath);
+          } catch (IOException ex) {
+            throw new CommitFailedException(
+                true /* retryable */,
+                false /* conflict */,
+                "Failed to write staged commit file due to: " + ex.getMessage(),
+                ex);
+          }
+        });
+  }
+
+  private void commitToUC(CommitMetadata commitMetadata, FileStatus kernelFileStatus)
+      throws CommitFailedException {
+    OperationTimer.timeOperation(
+        logger,
+        "Commit to UC",
+        ucTableId,
+        () -> {
+          final boolean isDisown =
+              commitMetadata.getCommitType()
+                  == CommitMetadata.CommitType.CATALOG_DOWNGRADE_TO_FILESYSTEM;
+
+          try {
+            ucClient.commit(
+                ucTableId,
+                tablePath.toUri(),
+                Optional.of(getCommitStructPayload(commitMetadata, kernelFileStatus)),
+                Optional.empty() /* lastKnownBackfilledVersion */,
+                isDisown,
+                Optional.empty() /* newMetadata */,
+                Optional.empty());
+            return null;
+          } catch (io.delta.storage.commit.CommitFailedException cfe) {
+            throw storageCFEtoKernelCFE(cfe);
+          } catch (IOException ex) {
```

Comment:
this won't suffer for issues on retryability because we have txnId that is checked someplace?

Reply:
Great question/callout. `txnId` is _not_ checked today -- we should do that in the future 👍 

But Kernel txn.commit loop already handles the case of:
- we try to commit 007.json
- IO exception
- we try again, see 007.json already exists
- we do NOT try to commit 008.json -- we throw https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/CommitStateUnknownException.java